### PR TITLE
Enable Python workflow with `BytePairTokenizer` and `StartEndPacker`

### PR DESCRIPTION
**Pull request by james77777778**
## Description of the change

On second thought, I think merging the Python workflow functionality into the original `BytePairTokenizer` and `StartEndPacker` layers will provide a more seamless UX. Users will benefit from a default Python workflow while maintaining a smooth experience when using these layers within `tf.data`.

Extensive tests have been added for these core modules:
- `keras_hub/src/tokenizers/byte_tokenizer_test.py`
- `keras_hub/src/layers/preprocessing/start_end_packer_test.py`

The diff is quite large due to updated tests. I found that all `BytePairTokenizer` tests were WRONG: BPE requires all elements in `merges` to be present in `vocabulary`.
Our `tf` and `tf-text` implementation is more permissive than widely adopted libraries like `tokenizers`.

Key updates for reviewers:
- Added `_allow_python_workflow` to `PreprocessingLayer` to indicate whether a layer supports Python workflow
- keras_hub/src/tokenizers/byte_pair_tokenizer.py
- keras_hub/src/tokenizers/byte_pair_tokenizer_test.py
- keras_hub/src/layers/preprocessing/start_end_packer.py
- keras_hub/src/layers/preprocessing/start_end_packer_test.py
- keras_hub/src/models/clip/clip_tokenizer.py: CLIPTokenizer has very different settings as a BPE tokenizer.
- keras_hub/src/models/clip/clip_tokenizer_test.py

## Colab Notebook

https://colab.research.google.com/drive/1YDHUwdjwgO0BGuelYLTRNvVB1Ulc1K7X?usp=sharing

image

## Checklist

- [x] I have added all the necessary unit tests for my change.
- [x] I have verified that my change does not break existing code and works with all backends (TensorFlow, JAX, and PyTorch).
- [x] My PR is based on the latest changes of the main branch (if unsure, rebase the code).
- [x] I have followed the Keras Hub [Model contribution guidelines](https://github.com/keras-team/keras-hub/blob/master/CONTRIBUTING_MODELS.md) in making these changes.
- [x] I have followed the Keras Hub [API design guidelines](https://github.com/keras-team/keras-hub/blob/master/API_DESIGN_GUIDE.md) in making these changes.
- [x] I have signed the [Contributor License Agreement](https://cla.developers.google.com/about).

### CODE REVIEW

**Comment by divyashreepathihalli** [comment]
@james77777778 can you please address the above comments?

File: keras_hub/src/layers/preprocessing/start_end_packer.py [hunk 1]
```diff
@@ -203,6 +216,152 @@ def call(
...
+        def _canonicalize_inputs(inputs):
+            if isinstance(inputs, (tuple, list)):
+                inputs = keras.tree.map_structure(convert_to_list, inputs)
+                if inputs and isinstance(inputs[0], (tuple, list)):
+                    return inputs, True
+                else:
+                    return [inputs], False
+            elif tf is not None and isinstance(
+                inputs, (tf.Tensor, tf.RaggedTensor)
+            ):
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_hub/src/layers/preprocessing/start_end_packer.py:257]: This is a false alarm. We are importing it like this:

```python
try:
    import tensorflow as tf
except ImportError:
    tf = None
```

File: keras_hub/src/tokenizers/byte_pair_tokenizer.py [hunk 2]
```diff
@@ -617,12 +767,71 @@ def process_unseen_tokens():
...
+                return [inputs], False
+            elif isinstance(inputs, (tuple, list)):
+                if not all(isinstance(i, str) for i in inputs):
+                    raise ValueError(
+                        "If a list or tuple is provided as input, all elements "
+                        "must be strings. "
+                        f"Received: {inputs}"
+                    )
+                return list(inputs), True
+            elif tf is not None and isinstance(inputs, tf.Tensor):
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_hub/src/tokenizers/byte_pair_tokenizer.py:786]: This is a false alarm. We are importing it like this:

```python
try:
    import tensorflow as tf
except ImportError:
    tf = None
```

File: keras_hub/src/utils/tensor_utils.py [hunk 3]
```diff
@@ -215,6 +215,46 @@ def convert(x):
...
+    ```
+    """
+    if in_no_convert_scope():
+        return x
+
+    def convert(x):
+        if x is None:
+            return x
+        if isinstance(x, str):
+            return tensor_to_list(x)
```
[the remark this answers is not in the document]
Comment (james77777778) [reply keras_hub/src/utils/tensor_utils.py]: Updated. Thanks!

**Comment by james77777778** [comment]
> @james77777778 can you please address the above comments?

All fixed. Let me know if anything is missing.

IMO, we should prioritize this PR since we're adding more models. Once the PR is merged, it will require developers to add a python workflow as well.

**Comment by divyashreepathihalli** [comment]
@james77777778 can you run gemini review a couple of times after your changes  just to make sure it catches all the initial set up bugs/issues.

File: keras_hub/src/layers/preprocessing/start_end_packer.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/layers/preprocessing/start_end_packer.py:257]: False flagging

File: keras_hub/src/tokenizers/byte_pair_tokenizer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/tokenizers/byte_pair_tokenizer.py:786]: False flagging

File: keras_hub/src/tokenizers/byte_pair_tokenizer.py [hunk 4]
```diff
@@ -10,12 +10,19 @@
 from typing import Iterable
 
 import keras
+import numpy as np
 import regex as re
+import tokenizers
 from keras.src.saving import serialization_lib
+from tokenizers import decoders
+from tokenizers import models
+from tokenizers import pre_tokenizers
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/tokenizers/byte_pair_tokenizer.py:19]: tokenizers is part of core dependency package.. users will have it with Keras Hub package installed.

File: keras_hub/src/models/clip/clip_tokenizer.py [hunk 5]
```diff
@@ -1,8 +1,15 @@
+import tokenizers
+from tokenizers import decoders
+from tokenizers import models
+from tokenizers import normalizers
+from tokenizers import pre_tokenizers
+from tokenizers import processors
```
[the remark this answers is not in the document]
Comment (sachinprasadhs) [reply keras_hub/src/models/clip/clip_tokenizer.py:6]: tokenizers is part of core dependency package.. users will have it with Keras Hub package installed.

File: keras_hub/src/layers/preprocessing/start_end_packer.py [hunk 6]
```diff
@@ -203,6 +216,152 @@ def call(
...
+                unbatched = inputs.shape.rank == 1
+                if unbatched:
+                    inputs = tf.expand_dims(inputs, 0)
+                if isinstance(inputs, tf.Tensor):
+                    inputs = convert_to_list(inputs)
+                else:
+                    inputs = inputs.to_list()
```
Comment (sachinprasadhs) [inline keras_hub/src/layers/preprocessing/start_end_packer.py:264]: I guess this won't create much of performance issue for small batch, do we have any observation on large batch on how much performance difference it would make?

**Review by sachinprasadhs — APPROVED** [review]
Thanks, this looks good to me. 

- I have marked the Gemini comments as resolved since they were not valid. 
- The token ID changes are due to the dynamic construction of the vocabulary, so I assume these updates are verified for all models. 
- I also left a comment asking about the performance difference on large batches of Ragged Tensors. If there is a difference, maybe we can add a note.

File: keras_hub/src/layers/preprocessing/start_end_packer.py (hunk 6 above)
Comment (james77777778) [reply keras_hub/src/layers/preprocessing/start_end_packer.py:264]: I wrote a benchmark script for this and here is the result:

| Batch Size | Seq Length | _call_tf (ms) | _call_python (ms) | Py / TF Time Ratio |
| :--- | :--- | :--- | :--- | :--- |
| 1 | 256 | 3.3826 | 0.2942 | 0.09x |
| 1 | 512 | 3.2579 | 0.8520 | 0.26x |
| 1 | 1024 | 3.1793 | 0.3114 | 0.10x |
| 1 | 2048 | 3.1864 | 2.4536 | 0.77x |
| 16 | 256 | 3.2399 | 4.3285 | 1.34x |
| 16 | 512 | 3.2352 | 8.7083 | 2.69x |
| 16 | 1024 | 3.2298 | 14.9271 | 4.62x |
| 16 | 2048 | 3.3053 | 32.0950 | 9.71x |
| 64 | 256 | 3.2533 | 16.7609 | 5.15x |
| 64 | 512 | 3.3434 | 31.6561 | 9.47x |
| 64 | 1024 | 3.4316 | 64.9770 | 18.93x |
| 64 | 2048 | 3.6953 | 145.5433 | 39.39x |
| 128 | 256 | 3.3871 | 34.1017 | 10.07x |
| 128 | 512 | 3.4771 | 66.5220 | 19.13x |
| 128 | 1024 | 3.5901 | 124.6988 | 34.73x |
| 128 | 2048 | 4.4436 | 288.3607 | 64.89x |

Python calls are much slower when handling large input matrices. However, this overhead is only at ms level which is relatively small compared to the model forward pass.

I think we can speed this up in a follow-up.

<details>
<summary>benchmark.py</summary>

```python
import os
import time

import numpy as np
import tensorflow as tf

os.environ["KERAS_BACKEND"] = "jax"
from keras_hub.src.layers.preprocessing.start_end_packer import StartEndPacker


def run_benchmark():
    batch_sizes = [1, 16, 64, 128]
    seq_lengths = [256, 512, 1024, 2048]

    packer = StartEndPacker(
        sequence_length=10,
        start_value=1,
        end_value=2,
        pad_value=0,
    )

    print("-" * 80)
    print(
        f"{'Batch Size':<12} | {'Seq Length':<12} | {'_call_tf (ms)':<15} | "
        f"{'_call_python (ms)':<15} | {'Py / TF'}"
    )
    print("-" * 80)

    num_runs = 50

    for batch_size in batch_sizes:
        for seq_len in seq_lengths:
            packer.sequence_length = seq_len

            # Generate synthetic data with varying lengths
            python_inputs = [
                np.random.randint(
                    10, 1000, size=np.random.randint(1, seq_len + 1)
                ).tolist()
                for _ in range(batch_size)
            ]

            # TF inputs (RaggedTensor)
            tf_inputs = tf.ragged.constant(python_inputs)

            # Warmup
            _ = packer._call_tf(tf_inputs, sequence_length=seq_len)
            _ = packer._call_python(python_inputs, sequence_length=seq_len)

            # Benchmark _call_tf
            start_tf = time.time()
            for _ in range(num_runs):
                _ = packer._call_tf(tf_inputs, sequence_length=seq_len)
            tf_time = (time.time() - start_tf) / num_runs * 1000

            # Benchmark _call_python
            start_py = time.time()
            for _ in range(num_runs):
                _ = packer._call_python(python_inputs, sequence_length=seq_len)
            py_time = (time.time() - start_py) / num_runs * 1000

            speedup = py_time / tf_time if tf_time > 0 else 0.0

            print(
                f"{batch_size:<12} | {seq_len:<12} | {tf_time:<15.4f} | "
                f"{py_time:<15.4f} | {speedup:.2f}x"
            )


if __name__ == "__main__":
    run_benchmark()

```

</details>

Comment (james77777778) [reply keras_hub/src/layers/preprocessing/start_end_packer.py:264]: Add a fast path in `_canonicalize_inputs` and here is the new results:

| Batch Size | Seq Length | _call_tf (ms) | _call_python (ms) | Py / TF Time Ratio |
| :--- | :--- | :--- | :--- | :--- |
|1            | 256          | 3.1927          | 0.0657          | 0.02x|
|1            | 512          | 3.2040          | 0.0951          | 0.03x|
|1            | 1024         | 3.1784          | 0.1612          | 0.05x|
|1            | 2048         | 3.1667          | 0.2634          | 0.08x|
|16           | 256          | 3.1764          | 0.5219          | 0.16x|
|16           | 512          | 3.2433          | 1.0028          | 0.31x|
|16           | 1024         | 3.2401          | 2.7646          | 0.85x|
|16           | 2048         | 3.3033          | 5.1976          | 1.57x|
|64           | 256          | 3.2393          | 2.0557          | 0.63x|
|64           | 512          | 3.3011          | 5.3230          | 1.61x|
|64           | 1024         | 3.4475          | 9.7785          | 2.84x|
|64           | 2048         | 3.7694          | 20.5601         | 5.45x|
|128          | 256          | 3.3804          | 4.1382          | 1.22x|
|128          | 512          | 3.5177          | 8.3573          | 2.38x|
|128          | 1024         | 3.7052          | 21.0118         | 5.67x|
|128          | 2048         | 4.4261          | 40.2029         | 9.08x|

**Comment by james77777778** [comment]
@divyashreepathihalli 

Gemini gave a few false flaggings during the review likely because this PR is too large.

@sachinprasadhs 

I’ve added a fast path and added the notes in `StartAndPacker`. Performance has improved for large inputs.
The root cause was that `keras.tree.map_structure` is quite expensive.

File: keras_hub/src/layers/preprocessing/start_end_packer.py (hunk 6 above)
Comment (sachinprasadhs) [reply keras_hub/src/layers/preprocessing/start_end_packer.py:264]: This is much better results with improvement, Thank you!

**Comment by sachinprasadhs** [comment]
Thank you for this feature! Enabling the Python workflow for these models on all backends is a great addition.

**Comment by divyashreepathihalli** [comment]
@james77777778 Thank you so much for your amazing work as always!!!
@sachinprasadhs Thanks for the help with the review.
